tripal/tripal
# Provide validation for already-used field terms

# New Feature

### Closes #1932 

### Tripal Version: 4.x

## Description
Currently there is no mechanism to prevent you from adding a field to a content type, and using a CV term that happens to already be in use by another field. This PR adds a check to the validation when adding a field.

## Testing?
~~We would like to include non-chado fields in the test, but currently issue #1966 prevents that. We will make a small change to allow us to test with a generic tripal field.~~ Fixed with PR #1958
~~1. edit `tripal/src/Plugin/Field/FieldType/TripalIntegerTypeItem.php` and after line number 34 add this line
    `if (!$termIdSpace) { return; }`
    i.e.~~
```
     $termIdSpace = $storage_settings['termIdSpace'];
     $termAccession = $storage_settings['termAccession'];
+    if (!$termIdSpace) { return; }
```
2. Go to Tripal -> Page Structure -> Contact -> Manage fields
3. +Create a new field
4. Tripal Fields -> Continue
5. For Label enter "tripal integer" -> select "Tripal Integer Field Type" -> Continue
6. For the Controlled Vocabulary Term, enter any term you like, I used 'Sequence length (data:1249)' -> Save Settings
7. Now we test! Click "+Create a new field"
8. Chado Fields -> Continue
9. Enter a Label, like "test" or something more interesting.
10. Select something we can add to contact, I used "Chado Integer Field Type" -> Continue
11. For "Table Column" I used `contact_id`
12. Enter a Controlled Vocabulary Term that is already in use, try the one you created at step 6
13. Observe a validation error.
![issue1932-1](https://github.com/user-attachments/assets/0ba35ded-1eeb-4627-abde-3a0c907e2663)
![issue1932-2](https://github.com/user-attachments/assets/8186ce3e-b687-4933-82d1-c9997b815dca)
14. Try the other existing terms on the contact content type, all of these should fail validation: 
```
Communication Contact (NCIT:C47954)
description (schema:description)
name (schema:name)
Project (NCIT:C47885)
```
15. Try any other term, e.g. `DNA (SO:0000352)`, validation should pass, and the field should be added successfully.
![issue1932-3](https://github.com/user-attachments/assets/a65f44ad-0366-4ee9-af69-e0f12bc3ab39)
16. To test the other validation that I fixed, add another field (I added another project field) and enter something not of the correct format as a term, e.g. `xxx`. The validation should not allow this and should correctly highlight the field in red. You can test this part on 4.x and see that the validation catches this, but does not highlight the field in red.
![issue1932-4](https://github.com/user-attachments/assets/5857d55b-0e8a-42dc-8497-9059c07cdca8)
![issue1932-5](https://github.com/user-attachments/assets/3aeeaa25-9d67-4445-ab92-ca8ab53d6b8d)

This is the appearance on 4.x:
![issue1932-6](https://github.com/user-attachments/assets/a0931608-41ce-4d51-99a7-4fe449e585be)
